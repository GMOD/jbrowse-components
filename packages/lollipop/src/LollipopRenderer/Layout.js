import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { doesIntersect2 } from '@gmod/jbrowse-core/util/range'

export class FloatingLayout {
  constructor({ width }) {
    if (!width) throw new Error('width required to make a new FloatingLayout')
    this.width = width
  }

  items = []

  layout = new Map()

  layoutDirty = false

  add(uniqueId, anchorLocation, width, height, data) {
    this.items.push({ uniqueId, anchorLocation, width, height, data })
    this.layoutDirty = true
  }

  /**
   * @returns {Map} of uniqueId => {x,y,anchorLocation,width,height,data}
   */
  getLayout(configuration) {
    if (!this.layoutDirty) return this.layout
    if (!configuration) throw new Error('configuration object required')

    const minY = readConfObject(configuration, 'minStickLength')

    // sort them by score ascending, so higher scores will always end up
    // stacked last (toward the bottom)
    const sorted = this.items.sort((a, b) => a.data.score - b.data.score)

    // bump them
    let maxBottom = 0
    const layoutEntries = Array(sorted.length)
    for (let i = 0; i < sorted.length; i += 1) {
      const currentItem = sorted[i]
      const { anchorLocation, width, height } = currentItem
      const start = anchorLocation - width / 2
      const end = start + width
      let top = minY
      let bottom = top + height

      // figure out how far down to put it
      for (let j = 0; j < i; j += 1) {
        const [, previouslyLayedOutItem] = layoutEntries[j]
        const {
          x: prevStart,
          y: prevTop,
          width: prevWidth,
          height: prevHeight,
        } = previouslyLayedOutItem
        const prevEnd = prevStart + prevWidth
        const prevBottom = prevTop + prevHeight
        if (
          doesIntersect2(prevStart, prevEnd, start, end) &&
          doesIntersect2(prevTop, prevBottom, top, bottom)
        ) {
          // bump this one to the bottom of the previous one
          top = prevBottom
          bottom = top + height
          j = -1 // we need to check all of them again after bumping
        }
      }

      // record the entry and update the maxBottom
      layoutEntries[i] = [
        currentItem.uniqueId,
        { ...currentItem, x: start, y: top },
      ]
      if (bottom > maxBottom) maxBottom = bottom
    }

    // try to tile them left to right all at the same level
    // if they don't fit, try to alternate them on 2 levels, then 3
    this.totalHeight = maxBottom
    this.layout = new Map(layoutEntries)
    this.layoutDirty = false
    return this.layout
  }

  getTotalHeight() {
    if (this.layoutDirty) this.getLayout()
    return this.totalHeight
  }

  toJSON() {
    return { pairs: [...this.getLayout()], totalHeight: this.getTotalHeight() }
  }

  static fromJSON() {
    throw new Error('not supported')
  }
}

export class PrecomputedFloatingLayout {
  constructor({ pairs, totalHeight }) {
    this.layout = new Map(pairs)
    this.totalHeight = totalHeight
  }

  add(uniqueId) {
    if (!this.layout.has(uniqueId))
      throw new Error(`layout error, precomputed layout is missing ${uniqueId}`)
  }

  getLayout() {
    return this.layout
  }

  getTotalHeight() {
    return this.totalHeight
  }

  static fromJSON(json) {
    return new PrecomputedFloatingLayout(json)
  }
}
