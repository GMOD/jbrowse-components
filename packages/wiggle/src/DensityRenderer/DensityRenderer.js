import { readConfObject } from '@gmod/jbrowse-core/configuration'
import ServerSideRendererType from '@gmod/jbrowse-core/pluggableElementTypes/renderers/ServerSideRendererType'
import { bpToPx } from '@gmod/jbrowse-core/util'
import {
  createCanvas,
  createImageBitmap,
} from '@gmod/jbrowse-core/util/offscreenCanvasPonyfill'
import React from 'react'
import { getScale } from './util'

export class WiggleBaseRenderer extends ServerSideRendererType {
  async makeImageData(props) {
    const { height, region, bpPerPx, highResolutionScaling = 1 } = props
    const width = (region.end - region.start) / bpPerPx
    if (!(width > 0) || !(height > 0)) {
      return { height: 0, width: 0 }
    }
    const canvas = createCanvas(
      Math.ceil(width * highResolutionScaling),
      height * highResolutionScaling,
    )
    const ctx = canvas.getContext('2d')
    ctx.scale(highResolutionScaling, highResolutionScaling)
    this.draw(ctx, props)

    const imageData = await createImageBitmap(canvas)
    return { imageData, height, width }
  }

  draw(/* ctx, props */) {
    /* draw features to context given props */
  }

  async render(renderProps) {
    const { height, width, imageData } = await this.makeImageData(renderProps)
    const element = React.createElement(
      this.ReactComponent,
      { ...renderProps, height, width, imageData },
      null,
    )
    return { element, imageData, height, width }
  }
}

export default class extends WiggleBaseRenderer {
  draw(ctx, props) {
    const {
      features,
      region,
      bpPerPx,
      scaleOpts,
      height,
      config,
      horizontallyFlipped,
    } = props
    const pivot = readConfObject(config, 'bicolorPivot')
    const pivotValue = readConfObject(config, 'bicolorPivotValue')
    const negColor = readConfObject(config, 'negColor')
    const posColor = readConfObject(config, 'posColor')
    let colorCallback
    let colorScale
    if (readConfObject(config, 'color') === '#f0f') {
      // default color, use posColor/negColor instead
      colorScale =
        pivot !== 'none'
          ? getScale({
              ...scaleOpts,
              pivotValue,
              range: [negColor, 'white', posColor],
            })
          : getScale({ ...scaleOpts, range: ['white', posColor] })
      colorCallback = feature => colorScale(feature.get('score'))
    } else {
      colorCallback = feature => readConfObject(config, 'color', [feature])
    }

    for (const feature of features.values()) {
      const s = feature.get('start')
      const e = feature.get('end')
      let leftPx = bpToPx(s, region, bpPerPx, horizontallyFlipped)
      let rightPx = bpToPx(e, region, bpPerPx, horizontallyFlipped)
      if (horizontallyFlipped) {
        ;[leftPx, rightPx] = [rightPx, leftPx]
      }
      const w = rightPx - leftPx + 0.3 // fudge factor for subpixel rendering
      ctx.fillStyle = colorCallback(feature)
      ctx.fillRect(leftPx, 0, w, height)
    }
  }
}
