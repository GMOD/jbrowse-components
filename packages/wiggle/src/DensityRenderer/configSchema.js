import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { types } from 'mobx-state-tree'

export default ConfigurationSchema(
  'WiggleRenderer',
  {
    color: {
      type: 'color',
      description: 'the color of track, overrides posColor and negColor',
      defaultValue: '#f0f',
    },
    posColor: {
      type: 'color',
      description: 'the color to use when the score is positive',
      defaultValue: 'blue',
    },
    negColor: {
      type: 'color',
      description: 'the color to use when the score is negative',
      defaultValue: 'red',
    },
    highlightColor: {
      type: 'color',
      description: 'the color of highlights over the wiggle track',
      defaultValue: 'rgba(255,255,0,0.5)',
    },
    clipColor: {
      type: 'color',
      description: 'the color of the clipping marker',
      defaultValue: 'red',
    },
    renderType: {
      type: 'stringEnum',
      model: types.enumeration('Rendering type', ['xyplot', 'density', 'line']),
      description: 'The type of rendering for wiggle data to use',
      defaultValue: 'xyplot',
    },
    filled: {
      type: 'boolean',
      description: 'fill in histogram',
      defaultValue: true,
    },
    bicolorPivot: {
      type: 'stringEnum',
      model: types.enumeration('Scale type', [
        'numeric',
        'mean',
        'z_score',
        'none',
      ]),
      description: 'type of bicolor pivot',
      defaultValue: 'numeric',
    },
    bicolorPivotValue: {
      type: 'number',
      defaultValue: 0,
      description: 'value to use for bicolor pivot',
    },
    summaryScoreMode: {
      type: 'stringEnum',
      model: types.enumeration('Score type', ['max', 'min', 'avg', 'whiskers']),
      description:
        'choose whether to use max/min/average or whiskers which combines all three into the same rendering',
      defaultValue: 'whiskers',
    },
  },
  { explicitlyTyped: true },
)
