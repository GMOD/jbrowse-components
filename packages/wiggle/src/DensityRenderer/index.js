import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import ConfigSchema from './configSchema'

export { default as ReactComponent } from './components/WiggleRendering'
export { default } from './DensityRenderer'

export const configSchema = ConfigurationSchema(
  'DensityRenderer',
  {},
  { baseConfiguration: ConfigSchema, explicitlyTyped: true },
)
