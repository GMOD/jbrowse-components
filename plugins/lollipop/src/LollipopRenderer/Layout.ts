import { readConfObject } from '@jbrowse/core/configuration'
import { doesIntersect2 } from '@jbrowse/core/util/range'
import { AnyConfigurationModel } from '@jbrowse/core/configuration/configurationSchema'

interface LayoutItem {
  uniqueId: string
  anchorLocation: number
  width: number
  height: number
  data: { score: number }
}

type LayoutEntry = LayoutItem & { x: number; y: number }

type LayoutMap = Map<string, LayoutEntry>

export class FloatingLayout {
  width: number

  totalHeight = 0

  constructor({ width }: { width: number }) {
    if (!width) throw new Error('width required to make a new FloatingLayout')
    this.width = width
  }

  items: LayoutItem[] = []

  layout: LayoutMap = new Map()

  layoutDirty = false

  add(
    uniqueId: string,
    anchorLocation: number,
    width: number,
    height: number,
    data: { score: number },
  ) {
    this.items.push({ uniqueId, anchorLocation, width, height, data })
    this.layoutDirty = true
  }

  /**
   * @returns Map of `uniqueId => {x,y,anchorLocation,width,height,data}`
   */
  getLayout(configuration?: AnyConfigurationModel) {
    if (!this.layoutDirty) return this.layout
    if (!configuration) throw new Error('configuration object required')

    const minY = readConfObject(configuration, 'minStickLength')

    // sort them by score ascending, so higher scores will always end up
    // stacked last (toward the bottom)
    const sorted = this.items.sort((a, b) => a.data.score - b.data.score)

    // bump them
    let maxBottom = 0
    const layoutEntries: [string, LayoutEntry][] = Array(sorted.length)
    for (let i = 0; i < sorted.length; i += 1) {
      const currentItem = sorted[i]
      const { anchorLocation, width, height } = currentItem
      const start = anchorLocation - width / 2
      const end = start + width
      let top = minY
      let bottom = top + height

      // figure out how far down to put it
      for (let j = 0; j < i; j += 1) {
        const [, previouslyLayedOutItem] = layoutEntries[j]
        const {
          x: prevStart,
          y: prevTop,
          width: prevWidth,
          height: prevHeight,
        } = previouslyLayedOutItem
        const prevEnd = prevStart + prevWidth
        const prevBottom = prevTop + prevHeight
        if (
          doesIntersect2(prevStart, prevEnd, start, end) &&
          doesIntersect2(prevTop, prevBottom, top, bottom)
        ) {
          // bump this one to the bottom of the previous one
          top = prevBottom
          bottom = top + height
          j = -1 // we need to check all of them again after bumping
        }
      }

      // record the entry and update the maxBottom
      layoutEntries[i] = [
        currentItem.uniqueId,
        { ...currentItem, x: start, y: top },
      ]
      if (bottom > maxBottom) maxBottom = bottom
    }

    // try to tile them left to right all at the same level
    // if they don't fit, try to alternate them on 2 levels, then 3
    this.totalHeight = maxBottom
    this.layout = new Map(layoutEntries)
    this.layoutDirty = false
    return this.layout
  }

  getTotalHeight() {
    if (this.layoutDirty)
      throw new Error('getTotalHeight does not work when the layout is dirty.')
    return this.totalHeight
  }

  serializeRegion() {
    return this.toJSON()
  }

  toJSON() {
    if (this.layoutDirty)
      throw new Error('toJSON does not work when the layout is dirty.')
    return { pairs: [...this.getLayout()], totalHeight: this.getTotalHeight() }
  }

  static fromJSON() {
    throw new Error('not supported')
  }
}

export class PrecomputedFloatingLayout {
  layout: LayoutMap

  totalHeight: number

  constructor({
    pairs,
    totalHeight,
  }: {
    pairs: [string, LayoutEntry][]
    totalHeight: number
  }) {
    this.layout = new Map(pairs)
    this.totalHeight = totalHeight
  }

  add(uniqueId: string) {
    if (!this.layout.has(uniqueId))
      throw new Error(`layout error, precomputed layout is missing ${uniqueId}`)
  }

  getLayout() {
    return this.layout
  }

  getTotalHeight() {
    return this.totalHeight
  }

  static fromJSON(
    json: ConstructorParameters<typeof PrecomputedFloatingLayout>[0],
  ) {
    return new PrecomputedFloatingLayout(json)
  }
}
