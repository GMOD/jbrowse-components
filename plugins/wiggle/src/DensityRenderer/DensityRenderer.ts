import WiggleBaseRenderer from '../WiggleBaseRenderer'
import { drawDensity } from '../drawDensity'
import type { RenderArgsDeserializedWithFeatures } from '../WiggleBaseRenderer'

export default class DensityRenderer extends WiggleBaseRenderer {
  async draw(
    ctx: CanvasRenderingContext2D,
    props: RenderArgsDeserializedWithFeatures,
  ) {
    return drawDensity(ctx, props)
  }
}
