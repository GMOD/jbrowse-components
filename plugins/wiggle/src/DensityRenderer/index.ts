import {
  readConfObject,
  ConfigurationSchema,
} from '@jbrowse/core/configuration'
import { featureSpanPx } from '@jbrowse/core/util'
import { Feature } from '@jbrowse/core/util/simpleFeature'
import { getScale } from '../util'

import ConfigSchema from '../configSchema'
import WiggleBaseRenderer, {
  WiggleBaseRendererProps,
} from '../WiggleBaseRenderer'

export { default as ReactComponent } from '../WiggleRendering'

export default class extends WiggleBaseRenderer {
  draw(ctx: CanvasRenderingContext2D, props: WiggleBaseRendererProps) {
    const { features, regions, bpPerPx, scaleOpts, height, config } = props
    const [region] = regions
    const pivot = readConfObject(config, 'bicolorPivot')
    const pivotValue = readConfObject(config, 'bicolorPivotValue')
    const negColor = readConfObject(config, 'negColor')
    const posColor = readConfObject(config, 'posColor')
    let colorCallback
    let colorScale: ReturnType<typeof getScale>
    if (readConfObject(config, 'color') === '#f0f') {
      // default color, use posColor/negColor instead
      colorScale =
        pivot !== 'none'
          ? getScale({
              ...scaleOpts,
              pivotValue,
              range: [negColor, 'white', posColor],
            })
          : getScale({ ...scaleOpts, range: ['white', posColor] })
      colorCallback = (feature: Feature) => colorScale(feature.get('score'))
    } else {
      colorCallback = (feature: Feature) =>
        readConfObject(config, 'color', [feature])
    }

    for (const feature of features.values()) {
      const [leftPx, rightPx] = featureSpanPx(feature, region, bpPerPx)
      const w = rightPx - leftPx + 0.3 // fudge factor for subpixel rendering
      ctx.fillStyle = colorCallback(feature)
      ctx.fillRect(leftPx, 0, w, height)
    }
  }
}
export const configSchema = ConfigurationSchema(
  'DensityRenderer',
  {},
  { baseConfiguration: ConfigSchema, explicitlyTyped: true },
)
