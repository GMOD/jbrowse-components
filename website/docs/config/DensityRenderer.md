---
id: densityrenderer
title: DensityRenderer
---

Note: this document is automatically generated from configuration objects in our
source code. See [Config guide](/docs/config_guide) for more info

### Source file

[plugins/wiggle/src/DensityRenderer/configSchema.ts](https://github.com/GMOD/jbrowse-components/blob/main/plugins/wiggle/src/DensityRenderer/configSchema.ts)

### DensityRenderer - Derives from

```js
baseConfiguration: baseWiggleRendererConfigSchema
```
