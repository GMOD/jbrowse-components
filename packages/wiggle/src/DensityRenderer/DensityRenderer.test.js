import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import Layout from '@gmod/jbrowse-core/util/layouts/GranularRectLayout'
import DensityRenderer, { configSchema, ReactComponent } from '.'

function DensityRendererPlugin() {
  return new DensityRenderer({
    name: 'DensityRenderer',
    ReactComponent,
    configSchema,
  })
}

test('empty', async () => {
  const result = await DensityRendererPlugin().makeImageData({
    layout: new Layout(),
    features: new Map(),
    region: {
      end: 100,
      start: 1,
      refName: 'ctgA',
      assemblyName: 'volvox',
    },
    scaleOpts: {
      domain: [0, 100],
      range: [0, 100],
      scaleType: 'linear',
    },
    config: {},
    bpPerPx: 3,
    highResolutionScaling: 1,
    horizontallyFlipped: false,
    height: 100,
    width: 800,
    blockKey: 'test',
  })
  expect(result).toEqual({ width: 0, height: 0 })
})

test('inverted mode and horizontally flipped', async () => {
  const result = await DensityRendererPlugin().makeImageData({
    layout: new Layout(),
    features: new Map([
      [
        't1',
        new SimpleFeature({ id: 't1', data: { start: 1, end: 100, score: 1 } }),
      ],
      [
        't2',
        new SimpleFeature({
          id: 't2',
          data: { start: 101, end: 200, score: 2 },
        }),
      ],
    ]),
    region: {
      end: 100,
      start: 1,
      refName: 'ctgA',
      assemblyName: 'volvox',
    },
    scaleOpts: {
      domain: [0, 100],
      range: [0, 100],
      scaleType: 'linear',
    },
    config: {},
    bpPerPx: 3,
    highResolutionScaling: 1,
    horizontallyFlipped: false,
    height: 100,
    width: 800,
    blockKey: 'test',
  })

  expect(result).toMatchSnapshot({
    imageData: expect.any(Object),
  })
})
