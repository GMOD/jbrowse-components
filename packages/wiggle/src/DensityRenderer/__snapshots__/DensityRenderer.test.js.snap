// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inverted mode and horizontally flipped 1`] = `
Object {
  "height": 100,
  "imageData": Any<Object>,
  "width": 33,
}
`;
